In [49]:
using Flux, MLDatasets, LinearAlgebra

In [50]:
using Flux: onehotbatch

In [51]:
using NPZ

In [52]:
# Load the dataset
dataset = npzread("data/bloodmnist.npz")

Dict{String, Array{UInt8}} with 6 entries:
  "train_labels" => [0x07; 0x03; … ; 0x06; 0x04;;]
  "test_labels"  => [0x03; 0x00; … ; 0x01; 0x07;;]
  "val_labels"   => [0x04; 0x00; … ; 0x01; 0x06;;]
  "test_images"  => [0xe8 0xe5 … 0xac 0xaf; 0xef 0xe3 … 0xb6 0xb8; … ; 0xe1 0xd…
  "val_images"   => [0xff 0xff … 0xc3 0xb9; 0xff 0xe7 … 0xfe 0xea; … ; 0xca 0xc…
  "train_images" => [0xfa 0xff … 0xd7 0xd9; 0xff 0xfd … 0xfe 0xfe; … ; 0xcd 0xc…

In [53]:
# Extract images and labels
imgs = dataset["train_images"]; @show imgs[1:4], typeof(imgs);

(imgs[1:4], typeof(imgs)) = (UInt8[0xfa, 0xff, 0xb9, 0xe2], Array{UInt8, 4})


In [57]:
labels = dataset["train_labels"]; @show lbls[1:4], typeof(lbls), length(labels);

(lbls[1:4], typeof(lbls), length(labels)) = (UInt8[0x07, 0x03, 0x06, 0x06], Matrix{UInt8}, 11959)


In [55]:
nTrain = length(imgs)

28127568

In [56]:
using BenchmarkTools

In [47]:
@btime trainData = vcat([hcat(float.(imgs[i])...)  for i in 1:nTrain]...);

  8.348 s (168764910 allocations: 5.03 GiB)


In [48]:
trainLabels = labels[1:nTrain]

LoadError: BoundsError: attempt to access 11959×1 Matrix{UInt8} at index [1:28127568]

# References
- [ ] [MedMNIST v2 - A large-scale lightweight benchmark
for 2D and 3D biomedical image classification](https://arxiv.org/pdf/2110.14795.pdf)
| Name | Source | Data  Modality | Task (# Classes / Labels) | # Samples | # Training / Validation / Test |
|-|-|-|-|-|-|
| BloodMNIST  | Acevedo et al. |  Blood Cell Microscope | MC (8) | 17,092  | 11,959 / 1,712 / 3,421 |